aiosmtpd/smtp.py

```diff
@@ -624,6 +631,38 @@ async def smtp_RSET(self, arg):
         status = await self._call_handler_hook('RSET')
         await self.push('250 OK' if status is MISSING else status)
 
+    def _parse_xclient_attributes(self, attributes):
+        """ Parse key-value arguments and return dict on success.
+        Raise ValueError on wrong arguments.
+        """
+        result = {}
+        for kv in attributes.split(' '):
+            try:
+                key, value = kv.split('=')
+            except ValueError as err:
+                raise ValueError('bad command parameter syntax') from err
+            if key not in XCLIENT_ATTRIBUTES:
+                raise ValueError(
+                    '{} is not a valid XCLIENT attribute'.format(key))
+            result[key] = value
+        return result
+
+    def _is_handler_has_xclient_hook(self):
+        return hasattr(self.event_handler, 'handle_XCLIENT')
+
+    @syntax('XCLIENT [key1=value1] ... [keyN=valueN]')
+    async def smtp_XCLIENT(self, arg):
+        if not self._is_handler_has_xclient_hook():
+            await self.push('550 feature not enabled')
+            return
+        try:
+            attributes = self._parse_xclient_attributes(arg)
```

Comment:
I much prefer doing the `XCLIENT` attribute validation here in this method, rather than having a method solely dedicated to performing XCLIENT attribute validation.